### PR TITLE
Integrating vmdkops_admin cli commands with esxcli - Part 1

### PR DESCRIPTION
**Pull request by pshahzeb**
## Description

This PR introduces xml based extension to integrate vmdkops admin cli commands with esxcli.
The xml file introduces vdvs namespace and furthermore nested namespace parallel to admin cli commandset. The experience is exactly the same (see manual test results below).
Essentially, each command fired through esxcli is mapped to a vmdkops_admin command. Input parameters and output formatting is specified with this.

Esxcli requires the actual command to generate output in xml format, hence the introduction of additional flag to specify output-format. Default is still plaintext.

The xml based extension is copied to appropriate location (required by esxcli) during the deploy-esx target. After VIB installation, admin can directly start using vdvs commands through esxcli

Note: currently I have used "vdvs". It's small and sweet but needs to be changed. Please suggest a better name.

## Testing
As of now, I have done manual testing of all the commands.
I plan to do automated tests and documentation update as separate PR.
```


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs volume ls
Volume      Datastore     VMGroup   Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By    Created Date
----------  ------------  --------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  ------------  ------------------------
test_vol_1  sharedVmfs-0  _DEFAULT  100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Thu Aug 10 07:25:08 2017
test_vol_2  sharedVmfs-0  _DEFAULT  100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.0  Thu Aug 10 07:25:12 2017

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs volume set --volume=test_vol_1@sharedVmfs-0 --vmgroup=_DEFAULT --options="access=read-only"
Successfully updated settings for test_vol_1@sharedVmfs-0

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs policy create --name some-policy --content '(("proportionalCapacity" i0)("hostFailuresToTolerate" i0))'
Successfully created policy: some-policy

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs policy ls
Policy Name  Policy Content                                              Active
-----------  ----------------------------------------------------------  ------
some-policy  (("proportionalCapacity" i0)("hostFailuresToTolerate" i0))  Unused


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs policy update --name some-policy --content '(("proportionalCapacity" i1))'
Successfully updated policy some-policy

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs policy rm --name=some-policy
Successfully removed policy: some-policy

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup ls
Uuid                                  Name      Description                  Default_datastore  VM_list
------------------------------------  --------  ---------------------------  -----------------  -------
11111111-1111-1111-1111-111111111111  _DEFAULT  This is the default vmgroup  _VM_DS

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup create --name=vmgroup1 --default-datastore=vsandatastore --vm-list=ubuntu-VM1.0
vmgroup 'vmgroup1' is created. Do not forget to run 'vmgroup vm add' to add vm to vmgroup.

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup update --name=vmgroup1 --description="New description of vmgroup1" --new-name=new-vmgroup1 --default-datastore=sharedVmfs-1
vmgroup modify succeeded

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup vm add --name=new-vmgroup1 --vm-list=ubuntu-VM0.0
vmgroup vm add succeeded

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup vm ls --name=new-vmgroup1
Uuid                                  Name
------------------------------------  ------------
564dfd1e-5a41-7007-6a8c-2936c51542c2  ubuntu-VM1.0
564d93b9-6908-6888-0514-fa7571b2e095  ubuntu-VM0.0

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup vm rm --name=new-vmgroup1 --vm-list=ubuntu-VM0.0
vmgroup vm rm succeeded


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup vm replace --name=new-vmgroup1 --vm-list=ubuntu-VM1.0
VM 'ubuntu-VM1.0' has already been associated with tenant 'new-vmgroup1', can't add it
[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup vm replace --name=new-vmgroup1 --vm-list=ubuntu-VM0.0
vmgroup vm replace succeeded



[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup access ls --name new-vmgroup1
Datastore      Allow_create  Max_volume_size  Total_size
-------------  ------------  ---------------  ----------
vsandatastore  True          Unset            Unset

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup access add --name=new-vmgroup1 --datastore=sharedVmfs-0  --volume-maxsize=500MB --volume-totalsize=1GB
vmgroup access add succeeded

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup access rm --name=new-vmgroup1 --datastore=sharedVmfs-0
vmgroup access rm succeeded

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs vmgroup access set --name=new-vmgroup1 --datastore=vsandatastore  --allow-create=True  --volume-maxsize=1000MB  --volume-totalsize=2GB
vmgroup access set succeeded


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs config status
DB_Mode: SingleNode (local DB exists)
DB_SharedLocation: N/A
DB_LocalPath: /etc/vmware/vmdkops/auth-db

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs config rm --local --confirm
Moved /etc/vmware/vmdkops/auth-db to backup file /etc/vmware/vmdkops/auth-db.bak_Thu_Aug_10_08:05:20_2017


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs config init --local
Warning: this feature is EXPERIMENTAL
Creating new DB at /etc/vmware/vmdkops/auth-db
Warning: Local configuration will not survive ESXi reboot. See KB2043564 for details


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs config init --datastore=vsandatastore
Found Config DB on other datastores. . Add '--force' flag to force the request execution. Additional information: [('local.0-0', '/vmfs/volumes/local.0-0/dockvols/vmdkops_config.db')]
[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs config init --datastore=vsandatastore --force
Warning: this feature is EXPERIMENTAL
Creating new DB at /vmfs/volumes/vsandatastore/dockvols/vmdkops_config.db
Creating a symlink to /vmfs/volumes/vsandatastore/dockvols/vmdkops_config.db at /etc/vmware/vmdkops/auth-db
Updating /etc/rc.local.d/local.sh

[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli vdvs config rm --unlink --confirm
Removed link /etc/vmware/vmdkops/auth-db
Updating /etc/rc.local.d/local.sh
```

## The help sections of the commands look like below:
```
root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs —help
Usage: esxcli storage vdvs {cmd} [cmd options]

Available Namespaces:
  vmgroup               Administer and monitor volume access control
  config                Init and manage Config DB to enable quotas and access control [EXPERIMENTAL]
  policy                Configure and display storage policy information
  volume                Manage vDVS volumes

Available Commands:
  status                Status of vdvs service


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs volume —help
Usage: esxcli storage vdvs volume {cmd} [cmd options]

Available Commands:
  ls                    List volumes
  set                   Edit settings for a given volume


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs volume ls --help
Usage: esxcli storage vdvs volume ls [cmd options]

Description:
  ls                    List volumes

Cmd options:
  --vmgroup=<str>       Displays volumes for a given vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs volume set --help
Usage: esxcli storage vdvs volume set [cmd options]

Description:
  set                   Edit settings for a given volume

Cmd options:
  --options=<str>       Options (specifically, access) to be set on the volume (required)
  --vmgroup=<str>       Name of the vmgroup the volume belongs to (required)
  --volume=<str>        Volume to set options for, specified as "volume@datastore" (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs volume  --help
Usage: esxcli storage vdvs volume {cmd} [cmd options]

Available Commands:
  ls                    List volumes
  set                   Edit settings for a given volume


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs policy  --help
Usage: esxcli storage vdvs policy {cmd} [cmd options]

Available Commands:
  create                Create a storage policy
  ls                    List storage policies and volumes using those policies
  rm                    Remove a storage policy
  update                Update the definition of a storage policy and all VSAN objects using that policy


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs policy ls --help
Usage: esxcli storage vdvs policy ls [cmd options]

Description:
  ls                    List storage policies and volumes using those policies

Cmd options:


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs policy create --help
Usage: esxcli storage vdvs policy create [cmd options]

Description:
  create                Create a storage policy

Cmd options:
  --content=<str>       The VSAN policy string (required)
  --name=<str>          The name of the policy (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs policy rm --help
Usage: esxcli storage vdvs policy rm [cmd options]

Description:
  rm                    Remove a storage policy

Cmd options:
  --name=<str>          The name of the policy (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs config init --help
Usage: esxcli storage vdvs config init [cmd options]

Description:
  init                  Init and manage Config DB to enable quotas and access control [EXPERIMENTAL]

Cmd options:
  --datastore=<str>     Config file will be placed on a datastore
  --force               Force operation, ignore warnings
  --local               Allows local (SingleNode) Init


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs config rm --help
Usage: esxcli storage vdvs config rm [cmd options]

Description:
  rm                    Init and manage Config DB to enable quotas and access control [EXPERIMENTAL]

Cmd options:
  --confirm             Explicitly confirm the operation
  --local               Remove only local link or local DB
  --no-backup           Do not create DB backup before removing
  --unlink              Remove the local link to shared DB


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs config mv --help
Usage: esxcli storage vdvs config mv [cmd options]

Description:
  mv                    Relocate config file from its current location [NOT SUPPORTED YET]

Cmd options:
  --force               Force operation, ignore warnings
  --to=<str>            Allows local (SingleNode) Init (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs config status --help
Usage: esxcli storage vdvs config status [cmd options]

Description:
  status                Show the status of the Config DB

Cmd options:


root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup --help
Usage: esxcli storage vdvs vmgroup {cmd} [cmd options]

Available Namespaces:
  access                Add or remove Datastore access and quotas for a vmgroup
  vm                    Add, removes, replaces and lists VMs in a vmgroup

Available Commands:
  create                Create a new vmgroup
  ls                    List vmgroups and the VMs they are applied to
  rm                    Delete a vmgroup
  update                Update a vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup create --help
Usage: esxcli storage vdvs vmgroup create [cmd options]

Description:
  create                Create a new vmgroup

Cmd options:
  --default-datastore=<str>
                        Datastore to be used by default for volumes placement (required)
  --description=<str>   The description of the vmgroup
  --name=<str>          The name of the vmgroup (required)
  --vm-list=<str>       A list of VM names to place in this vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup ls --help
Usage: esxcli storage vdvs vmgroup ls [cmd options]

Description:
  ls                    List vmgroups and the VMs they are applied to

Cmd options:


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup rm --help
Usage: esxcli storage vdvs vmgroup rm [cmd options]

Description:
  rm                    Delete a vmgroup

Cmd options:
  --name=<str>          The name of the vmgroup (required)
  --remove-volumes      BE CAREFUL: Removes this vmgroup volumes when removing a vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup update --help
Usage: esxcli storage vdvs vmgroup update [cmd options]

Description:
  update                Update a vmgroup

Cmd options:
  --default-datastore=<str>
                        Datastore to be used by default for volumes placement
  --description=<str>   The new description of the vmgroup
  --name=<str>          The name of the vmgroup (required)
  --new-name=<str>      The new name of the vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup vm --help
Usage: esxcli storage vdvs vmgroup vm {cmd} [cmd options]

Available Commands:
  add                   Add VM(s) to a vmgroup
  ls                    list VMs in a vmgroup
  replace               Replace VM(s) for a vmgroup
  rm                    Remove VM(s) from a vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup vm add --help
Usage: esxcli storage vdvs vmgroup vm add [cmd options]

Description:
  add                   Add VM(s) to a vmgroup

Cmd options:
  --name=<str>          The name of the vmgroup (required)
  --vm-list=<str>       A list of VM names to place in this vmgroup (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup vm ls --help
Usage: esxcli storage vdvs vmgroup vm ls [cmd options]

Description:
  ls                    list VMs in a vmgroup

Cmd options:
  --name=<str>          Vmgroup name (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup vm rm --help
Usage: esxcli storage vdvs vmgroup vm rm [cmd options]

Description:
  rm                    Remove VM(s) from a vmgroup

Cmd options:
  --name=<str>          Vmgroup to remove the VM from (required)
  --vm-list=<str>       A list of VM names to rm from this vmgroup (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup vm replace --help
Usage: esxcli storage vdvs vmgroup vm replace [cmd options]

Description:
  replace               Replace VM(s) for a vmgroup

Cmd options:
  --name=<str>          Vmgroup to replace the VM for (required)
  --vm-list=<str>       A list of VM names to replace for this vmgroup (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup access --help
Usage: esxcli storage vdvs vmgroup access {cmd} [cmd options]

Available Commands:
  add                   Add a datastore access for a vmgroup
  ls                    List access for a vmgroup
  rm                    Remove datastore access for a vmgroup
  set                   Modify datastore access for a vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup access add --help
Usage: esxcli storage vdvs vmgroup access add [cmd options]

Description:
  add                   Add a datastore access for a vmgroup

Cmd options:
  --allow-create        Allow create and delete on datastore if set
  --datastore=<str>     Datastore name (required)
  --name=<str>          The name of the vmgroup (required)
  --volume-maxsize=<str>
                        Maximum size of the volume that can be created
  --volume-totalsize=<str>
                        Maximum total size of all volume that can be created on the datastore for this vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup access ls --help
Usage: esxcli storage vdvs vmgroup access ls [cmd options]

Description:
  ls                    List access for a vmgroup

Cmd options:
  --name=<str>          Vmgroup name (required)


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup access set --help
Usage: esxcli storage vdvs vmgroup access set [cmd options]

Description:
  set                   Modify datastore access for a vmgroup

Cmd options:
  --allow-create        Allow create and delete on datastore if set
  --datastore=<str>     Datastore name (required)
  --name=<str>          The name of the vmgroup (required)
  --volume-maxsize=<str>
                        Maximum size of the volume that can be created
  --volume-totalsize=<str>
                        Maximum total size of all volume that can be created on the datastore for this vmgroup


[root@sc2-rdops-vm03-dhcp-118-213:~] esxcli storage vdvs vmgroup access rm --help
Usage: esxcli storage vdvs vmgroup access rm [cmd options]

Description:
  rm                    Remove datastore access for a vmgroup

Cmd options:
  --datastore=<str>     Datastore name (required)
  --name=<str>          The name of the vmgroup (required)
```

### CODE REVIEW

**Comment by pshahzeb** [comment]
volume ls -c flag isn't supported as of now
Because for table-like output,  -c flag specifies column names while each time a command is run.
In case of esxcli  we need to specify which are the columns to be displayed for a command through xml. This is can be set once, when vmdkops_admin cli extension is read from xml file.

File: esx_service/cli/vmdkops_admin.xml [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+        <namespace path="vdvs.config">
+            <description>Init and manage Config DB to enable quotas and access control [EXPERIMENTAL]</description>
+        </namespace>
+    </namespaces>
+    <commands>
+        <!-- Volume commands -->
+        <command path="vdvs.volume.ls">
+            <description>'List volumes'</description>
+            <input-spec>
+                <parameter name="vmgroup" type="string" required="false">
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.xml]: Can vdvs.volume.ls specify --vmgroup option to only show volumes belong to a specific vmgroup? If yes, please add test for that.

File: esx_service/cli/vmdkops_admin.xml [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+            <output-spec>
+                <string/>
+            </output-spec>
+            <format-parameters>
+                <formatter>simple</formatter>
+            </format-parameters>
+            <execute>/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py --output-format=xml policy rm --name=$val{name}</execute>
+        </command>
+        <command path="vdvs.policy.ls">
+            <description>List storage policies and volumes using those policies</description>
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.xml]: Does vdvs.policy.ls list volumes using those policies?

File: esx_service/cli/vmdkops_admin.xml [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+            </output-spec>
+            <has-updates value="true"/>
+            <format-parameters>
+                <formatter>table</formatter>
+                <format-parameter name="fields:vdvs">Policy Name,Policy Content,Active</format-parameter>
+            </format-parameters>
+            <execute>/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py --output-format=xml policy ls</execute>
+        </command>
+        <command path="vdvs.policy.update">
+            <description>'Update the definition of a storage policy and all VSAN objects using that policy'</description>
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.xml]: What do you mean "vdvs.policy.update" also update all VSAN objects using that policy?

File: esx_service/cli/vmdkops_admin.xml [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+            <has-updates value="true"/>
+            <format-parameters>
+                <formatter>table</formatter>
+                <format-parameter name="fields:vdvs">Uuid, Name, Description, Default_datastore, VM_list</format-parameter>
+            </format-parameters>
+            <execute>/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py --output-format=xml vmgroup ls</execute>
+        </command>
+        <!-- vmgroup vm commands -->
+        <command path="vdvs.vmgroup.vm.add">
+            <description>Add VM(s)  to a vmgroup</description>
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.xml]: Nit: one extra space after VM(s)

File: esx_service/cli/vmdkops_admin.xml [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+            <description>Modify datastore access for a vmgroup</description>
+            <input-spec>
+                <parameter name="name" type="string" required="true">
+                    <description>The name of the vmgroup</description>
+                </parameter>
+                <parameter name="datastore" type="string" required="true">
+                    <description>Datastore name</description>
+                </parameter>
+                <parameter name="allow-create" type="bool" required="false">
+                    <description>Allow create and delete on datastore if set</description>
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.xml]: "Allow create and delete on datastore if set to true". This value can also be set to "false" by user.

File: esx_service/cli/vmdkops_admin.xml [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+                <string />
+            </output-spec>
+            <has-updates value="true"/>
+            <format-parameters>
+                <formatter>simple</formatter>
+            </format-parameters>
+            <execute>/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py --output-format=xml vmgroup access set --name=$val{name} --datastore=$val{datastore} $if{allow-create, --allow-create=$val{allow-create}} $if{volume-maxsize, --volume-maxsize=$val{volume-maxsize}} $if{volume-totalsize, --volume-totalsize=$val{volume-totalsize}} </execute>
+        </command>
+        <command path="vdvs.vmgroup.access.rm">
+            <description>Modify datastore access for a vmgroup</description>
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.xml]: Should be "Remove datastore access for a vmgroup", looks like a copy/paste issue :)

File: esx_service/cli/vmdkops_admin.xml [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+                <string />
+            </output-spec>
+            <has-updates value="true"/>
+            <format-parameters>
+                <formatter>simple</formatter>
+            </format-parameters>
+            <execute>/usr/lib/vmware/vmdkops/bin/vmdkops_admin.py --output-format=xml vmgroup access rm --name=$val{name} --datastore=$val{datastore}</execute>
+        </command>
+        <command path="vdvs.vmgroup.access.ls">
+            <description>list access for a vmgroup</description>
```
Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.xml]: Change to "List datastore access for a vmgroup"

File: esx_service/utils/cli_xml.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+def createMessage(message):
+    """
+    Generates xml with output tag with a string field for representing message
+    Returns formatted xml doc object
+    """
+    doc, output_elem = createDocXML()
+    str_elem = doc.createElement(STR_TAG)
+    output_elem.appendChild(str_elem)
+    str_elem.appendChild(doc.createTextNode(message))
+    return doc.toprettyxml()
```
Comment (lipingxue) [inline esx_service/utils/cli_xml.py]: Nit: Need to have a empty line after this line to make the end of the file

File: esx_service/vsan_policy.py [hunk 9]
```diff
@@ -204,12 +203,12 @@ def delete(name):
 
     vmdk = policy_in_use(path, name)
     if vmdk:
-        return 'Error: Cannot remove. Policy is in use by {0}'.format(vmdk)
+        return 'Cannot remove. Policy is in use by {0}'.format(vmdk)
```
Comment (lipingxue) [inline esx_service/vsan_policy.py]: Why "Error:" is removed?

File: esx_service/vsan_policy.py [hunk 10]
```diff
@@ -204,12 +203,12 @@ def delete(name):
...
     try:
         os.remove(policy_path(name))
     except:
         logging.exception("Failed to remove %s policy file", name)
-        return 'Error: {0} does not exist'.format(name)
+        return 'Policy {0} does not exist'.format(name)
```
Comment (lipingxue) [inline esx_service/vsan_policy.py]: Why "Error:" is removed?

File: esx_service/utils/cli_xml.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License
+
+# A small module for creating vmdkops admin cli command outputs in xml format
+
+import xml.dom.minidom
+
+# top level tag in output requires esxcli namespace declared
+ESXCLI_NS = "http://www.vmware.com/Products/ESX/6.0/esxcli/"
+
```
Comment (govint) [inline esx_service/utils/cli_xml.py]: Isn't 6.5 also supported? is it the same as 6.0?

File: misc/scripts/deploy-tools.sh [hunk 12]
```diff
@@ -210,6 +211,8 @@ function deployESXInstall {
 }
 
 function deployESXPost {
+    # restart hostd so that esxcli extension file is read
+    $SSH $TARGET $HOSTD restart
```
Comment (govint) [inline misc/scripts/deploy-tools.sh]: I'm not sure its ok to restart hostd when the VIB is installed. It may be disruptive. If this is the typical  practice then I guess thats fine. Just checking.

File: esx_service/cli/vmdkops_admin.xml [hunk 13]
```diff
@@ -0,0 +1,571 @@
+<?xml version="1.0"?>
+<plugin
+    xmlns="http://www.vmware.com/Products/ESX/6.0/esxcli/">
+    <!-- esxcli extension syntax version -->
+    <version>1.0.0</version>
+    <namespaces>
+        <namespace path="vdvs">
```
Comment (govint) [inline esx_service/cli/vmdkops_admin.xml]: As suggested "storage containers volumes ..." seems a good option. I'd suggest that all  options would be under "storage containers volumes". Because "storage containers policy" would be misleading as we don't deal with containers as such - only container volumes.

Comment (pshahzeb) [inline esx_service/cli/vmdkops_admin.xml]: Yes. Can use main namespace as ```esxcli storage containers <further commands>```.
However, putting everything under volumes also seems to be unsound.
For eg. 

the command to set access to a vmgroup would endup something like:
```esxcli storage containers volumes vmgroup access set <options>```

What do you think?

Comment (lipingxue) [inline esx_service/cli/vmdkops_admin.xml]: How about we use "container storage manager", under that, we can have "volume", "vmgroup", "policy" as subcommend.

Comment (govint) [inline esx_service/cli/vmdkops_admin.xml]: Or storage container-storage volumes/vmgroups/config/policies/etc/

File: esx_service/cli/vmdkops_admin.xml (hunk 1 above)
Comment (pshahzeb) [inline esx_service/cli/vmdkops_admin.xml]: Yes. I plan to do tests as a separate PR

File: esx_service/cli/vmdkops_admin.xml (hunk 2 above)
Comment (pshahzeb) [inline esx_service/cli/vmdkops_admin.xml]: Yes. Under Active column. 
```
Policy Name  Policy Content                                              Active
-----------  ----------------------------------------------------------  ------
bad_string   (("proportionalCapacity" i0) ("hostFailuresToTolerate" i3)  test2
```

File: esx_service/cli/vmdkops_admin.xml (hunk 3 above)
Comment (pshahzeb) [inline esx_service/cli/vmdkops_admin.xml]: It also updates vsan volumes that were created using this policy

File: esx_service/vsan_policy.py (hunk 9 above)
Comment (pshahzeb) [inline esx_service/vsan_policy.py]: Because this error message is prepended with 'ERROR' string when through vmdkops_admin.
This is common for every failed vmdkops admin command. So we don't need additional string.

File: esx_service/vsan_policy.py (hunk 10 above)
Comment (pshahzeb) [inline esx_service/vsan_policy.py]: See reply above.

File: esx_service/utils/cli_xml.py (hunk 11 above)
Comment (pshahzeb) [inline esx_service/utils/cli_xml.py]: Yes. Tested.

File: misc/scripts/deploy-tools.sh (hunk 12 above)
Comment (pshahzeb) [inline misc/scripts/deploy-tools.sh]: Couldn't come across a better way to do this. This is what the esxcli doc mentions about discovering extensions
```
In ESX 5.0 this step is performed once by the hostagent daemon when it starts up. Any commands available to the system during hostd startup will be registered as valid esxcli commands.
```

Ignore the 5.0. The documentation is old.

I ll check again if we can do anything better than this.

File: esx_service/cli/vmdkops_admin.py [hunk 14]
```diff
@@ -1205,15 +1228,16 @@ def err_out(_msg, _info=None):
...
-        print("Additional information: {}".format(_info))
+        _msg = _msg + (". Additional information: {}".format(_info))
+    print(_msg)
     return _msg
 
 
 def err_override(_msg, _info):
     """A helper to print messates with extra help about --force flag"""
-    new_msg = "Error: {}".format(_msg) + " . Add '--force' flag to force the request execution"
+    new_msg = "{}".format(_msg) + " . Add '--force' flag to force the request execution"
```
Comment (ashahi1) [inline esx_service/cli/vmdkops_admin.py]: messates?

**Comment by pshahzeb** [comment]
@lipingxue @govint @ashahi1 
I have addressed the comments and updated the PR description to show how help looks like. PTAL

File: misc/scripts/deploy-tools.sh (hunk 12 above)
Comment (shuklanirdesh82) [inline misc/scripts/deploy-tools.sh]: better to be part of installESX (VIB installation)..

a question here: what will be the end user experience?

Comment (pshahzeb) [inline misc/scripts/deploy-tools.sh]: Yes. Agree.
Although I am not sure if we can specify hostd restart during VIB installation. 
Currently we do installation as localcli software vib install <vib_name>
Will be good to have it done during installation because it presently it will require manual hostd restart.

Comment (shuklanirdesh82) [inline misc/scripts/deploy-tools.sh]: 1. how about through `esxcli`? 
2. https://github.com/vmware/docker-volume-vsphere/pull/1735#discussion_r132777695 : it is a requirement of hostd restart then README and installation guideline should be updated as part of this review.

Comment (shuklanirdesh82) [inline misc/scripts/deploy-tools.sh]: As discussed offline .. it should be part of installESX routine; need to move there.
